experiments/MoarCSets/SInstances.jl

```diff
@@ -0,0 +1,406 @@
+
+function Base.getproperty(AD::Type{T},i::Symbol) where
+  {Ob,Hom,Dom,Codom,T<:CatDesc{Ob,Hom,Dom,Codom}}
+  @match i begin
+    :ob => Ob
+    :hom => Hom
+    :dom => Dom
+    :codom => Codom
+    _ => invoke(Base.getproperty, Tuple{Type,Symbol}, AD, i)
+  end
+end
+
+function cd_ob_num(CD::Type{T}, ob::Symbol) where {Ob,Hom,Dom,Codom, T <: CatDesc{Ob,Hom,Dom,Codom}}
+  findfirst(Ob .== ob)::Int
+end
+
+function cd_hom_num(CD::Type{T}, hom::Symbol) where {Ob,Hom,Dom,Codom, T <: CatDesc{Ob,Hom,Dom,Codom}}
+  findfirst(Hom .== hom)::Int
+end
+
+function cd_dom(CD::Type{T}, hom::Symbol) where {Ob,Hom,Dom,Codom, T <: CatDesc{Ob,Hom,Dom,Codom}}
+  Ob[Dom[cd_hom_num(CD,hom)]]
+end
+
+function cd_codom(CD::Type{T}, hom::Symbol) where {Ob,Hom,Dom,Codom, T <: CatDesc{Ob,Hom,Dom,Codom}}
+  Ob[Codom[cd_hom_num(CD,hom)]]
+end
+
+struct AttrDesc{CD,Data,Attr,ADom,ACodom}
+  function AttrDesc{CD,Data,Attr,ADom,ACodom}() where {CD,Data,Attr,ADom,ACodom}
+    new{CD,Data,Attr,ADom,ACodom}()
+  end
+  function AttrDesc(pres::Presentation{Schema})
+    CD = CatDescType(pres)
+    datas, attrs = generators(pres, :Data), generators(pres,:Attr)
+    data_syms, attr_syms = nameof.(datas), nameof.(attrs)
+    ob_num = ob -> findfirst(CD.ob .== ob)::Int
+    data_num = ob -> findfirst(data_syms .== ob)::Int
+    new{CD,Tuple(data_syms), Tuple(attr_syms),
+        Tuple(@. ob_num(nameof(dom(attrs)))), Tuple(@. data_num(nameof(codom(attrs))))}()
+  end
+  function AttrDesc(::CatDesc{Ob,Hom,Dom,Codom}) where {Ob,Hom,Dom,Codom}
+    new{CatDesc{Ob,Hom,Dom,Codom},(),(),(),()}
+  end
+end
+
+AttrDescType(pres::Presentation{Schema}) = typeof(AttrDesc(pres))
+
+function Base.getproperty(AD::Type{T}, i::Symbol) where
+  {CD,Data,Attr,ADom,ACodom,T <: AttrDesc{CD,Data,Attr,ADom,ACodom}}
+  @match i begin
+    :cd => CD
+    :data => Data
+    :attr => Attr
+    :adom => ADom
+    :acodom => ACodom
+    _ => invoke(Base.getproperty, Tuple{Type,Symbol}, AD, i)
+  end
+end
+
+function ad_data_num(AD::Type{T}, data::Symbol) where
+  {CD,Data,Attr,ADom,ACodom,T <: AttrDesc{CD,Data,Attr,ADom,ACodom}}
+  findfirst(Data .== data)::Int
+end
+
+function ad_attr_num(AD::Type{T}, attr::Symbol) where
+  {CD,Data,Attr,ADom,ACodom,T <: AttrDesc{CD,Data,Attr,ADom,ACodom}}
+  findfirst(Attr .== attr)::Int
+end
+
+function ad_dom(AD::Type{T}, attr::Symbol) where
+  {CD,Data,Attr,ADom,ACodom,T <: AttrDesc{CD,Data,Attr,ADom,ACodom}}
+  CD.ob[ADom[ad_attr_num(AD,attr)]]
+end
+
+function ad_codom(AD::Type{T}, attr::Symbol) where
+  {CD,Data,Attr,ADom,ACodom,T <: AttrDesc{CD,Data,Attr,ADom,ACodom}}
+  Data[ACodom[ad_attr_num(AD,attr)]]
+end
+
+function make_index(CD::Type{<:CatDesc},AD::Type{<:AttrDesc},Ts::Type{<:Tuple},Idxed::Tuple)
+    function make_idx(name)
+      if name ∈ CD.hom
+        Vector{Vector{Int}}()
+      elseif name ∈ AD.attr
+        Dict{Ts.parameters[ad_codom(AD,name)],Vector{Int}}()
+      else
+        error("invalid index parameter")
```

Comment:
A nicer error message would be something like `"$name is not a morphism or data attribute: cannot create index"`